jax/_src/nn/functions.py

```diff
@@ -1008,9 +1032,20 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
         mask_type = MaskType.CAUSAL
       elif query_seq_lengths is not None:
         mask_type = MaskType.PADDING
+      # CuDNN supports only the left window with an exclusive boundary with
+      # causal mask is enabled.
+      sliding_window = None
+      if local_window_size is not None:
+        l_window, r_window = local_window_size
+        if r_window == 0 or mask_type == MaskType.CAUSAL:
```

Comment:
If r_window != 0 and causal attention is used, then the r_window seems to be silently ignored. Shouldn't this fail instead and tell the user that the combination is not allowed?

Reply:
It will also fail when r_window != 0 and non causal masking is used. The comment above suggests this should be fine. Is that allowed?

Reply:
> If r_window != 0 and causal attention is used, then the r_window seems to be silently ignored. Shouldn't this fail instead and tell the user that the combination is not allowed?

No, it should work. This is intentionally, since the causal mask will effectively cover all the r_window.

> It will also fail when r_window != 0 and non causal masking is used. The comment above suggests this should be fine. Is that allowed?

Yes, this will fail. I think the comment above focuses on the behavior of cudnn, who only supports l_window + causal mask. And these branches try to translate the (l_window, r_window) + causal_mask to the cudnn with just one relaxation, i.e. when r_window is not 0 but causal is set, we are pretty sure the r_window could be safely ignored.